pallets/permissions/src/lib.rs

```diff
@@ -167,28 +185,26 @@ impl<T: Config> Pallet<T> {
 		with_role: T::Role,
 		location: T::Location,
 		role: T::Role,
-	) -> DispatchResult {
+	) -> Result<Who, DispatchError> {
 		if let Ok(who) = ensure_signed(origin.clone()) {
-			let has_role =
-				Permission::<T>::get(who.clone(), location.clone()).map_or(None, |roles| {
-					if roles.exists(role.clone()) {
-						Some(with_role)
-					} else {
-						None
-					}
-				});
+			let has_role = Permission::<T>::get(who.clone(), location.clone()).and_then(|roles| {
+				if roles.exists(with_role.clone()) {
+					Some(with_role)
+				} else {
+					None
+				}
+			});
 
 			if T::Editors::contains(&(who, has_role, location, role)) {
-				return Ok(());
+				return Ok(Who::Editor);
 			}
 		}
 
-		Self::ensure_admin(origin)
+		Self::ensure_admin(origin).map(|_| Who::Root)
```

Comment:
I think with the approach of the weights, we should early return here in any case. As a `Origin::signed(who)`, where `who` is not contained in the `Editors` will get the weight for the `Who::Root`, which is less and not correct. 

Or we move the admin check to the top.

Reply:
Nope, you are right. Then nevermind.

Reply:
And what if you are Admin but not editor? Can a signed extrinsic pass the admin check? Yes, right? Then this edge case would get the wrong weight.

Reply:
I have refactored it a lil bit. take another look ?